src/nupic/encoders/Scalar.cpp

```diff
@@ -0,0 +1,244 @@
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementations of the ScalarEncoder and PeriodicScalarEncoder
+ */
+
+#include <cstring> // memset
+#include <cmath>
+#include <nupic/encoders/Scalar.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic
+{
+  static Real64 extractScalar(const ArrayBase & input)
+  {
+    switch (input.getType())
+    {
+    case NTA_BasicType_Byte:
+      return ((Byte*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int16:
+      return ((Int16*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt16:
+      return ((UInt16*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int32:
+      return ((Int32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt32:
+      return ((UInt32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int64:
+      return ((Int64*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt64:
+      return ((UInt64*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real32:
+      return ((Real32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real64:
+      return ((Real64*) input.getBuffer())[0];
+      break;
+    default:
+      NTA_THROW << "extractScalar: Unsupported type " << input.getType();
+    }
+  }
+
+  ScalarEncoder::ScalarEncoder(
+    int w, double minValue, double maxValue, int n, double radius,
+    double resolution, bool clipInput)
+    :w_(w),
+     minValue_(minValue),
+     maxValue_(maxValue),
+     clipInput_(clipInput)
+  {
+    if ((n != 0 && (radius != 0 || resolution != 0)) ||
+        (radius != 0 && (n != 0 || resolution != 0)) ||
+        (resolution != 0 && (n != 0 || radius != 0)))
+    {
+      NTA_THROW <<
+        "Only one of n/radius/resolution can be specified for a ScalarEncoder.";
+    }
+
+    const int extentWidth = maxValue - minValue;
+    if (extentWidth <= 0)
+    {
+      NTA_THROW << "minValue must be < maxValue. minValue=" << minValue <<
+        " maxValue=" << maxValue;
+    }
+
+    if (n != 0)
+    {
+      n_ = n;
+
+      // Distribute nBuckets points along the domain [minValue, maxValue],
+      // including the endpoints. The resolution is the width of each band
+      // between the points.
+      const int nBuckets = n - (w - 1);
+      const int nBands = nBuckets - 1;
+      bucketsPerUnit_ = nBands / extentWidth;
```

Comment:
I find it hard to read with `bucketsPerUnit_`. As long as you validate the numbers are ok it would be fine to store `bucketSize_` or something like that that you divide by instead of the inverse?